minvws/nl-kat-coordination
# Fix rabbtimq exception handling in scheduler

### Changes

The following scheduler errors were encountered:

```
root-scheduler-1  | pika.exceptions.AMQPHeartbeatTimeout: No activity or too many missed heartbeats in the last 60 seconds 
root-scheduler-1  | pika.exceptions.ConnectionWrongStateError: BlockingConnection.add_callback_threadsafe() called on closed or closing connection.
root-scheduler-1  | RuntimeError: cannot schedule new futures after interpreter shutdown
```

We were expected that the scheduler would clean up after an unhandled exception. However, when trying to shutdown and clean up the listeners the scheduler throws another exception because the rabbitmq channel is closed. Which in turn is another unhandled exception triggering yet another shutdown. This results in a deadlock where the clean up of the whole scheduler application is stalled and other running threads keep running.

Within this PR we made sure we try-catch the `basic_consume` (start of consuming rabbitmq messaging), since we saw from the stack trace that the `AMQPHeartbeatTimeout:` was thrown there. Prior to this PR we didn't catch all exception that could've been throw by `basic_consume`, which resulted in unhandled exception we couldn't recover from.

Additionally, we made sure that when we shutdown a listener we also check whether we have an active connection and channel that we can close. This was the cause of the second unhandled exception.

### Issue link

Closes #2948 

### QA Instructions

Since the initial error was a lost heartbeat timeout, which could be hard to trigger, we can try to mimic this behaviour to start openkat, run a few jobs and terminate the rabbitmq instance to make sure the scheduler shut downs correctly.

```
$ docker-compose kill rabbitmq
```
